enhancements/storage/storage-hypershift.md

```diff
@@ -0,0 +1,1050 @@
+  - Operator X will degrade with message "Failed to launch webhook server" and reason "WehhookServerFailed".
+  - The metric `webhook_admission_duration_seconds("openpolicyagent-admission", "mutating", "put", "false")`
+    will show >1s latency and alert `WebhookAdmissionLatencyHigh` will fire.
+
+- disable the API extension (e.g. remove MutatingWebhookConfiguration `xyz`, remove APIService `foo`)
+
+  - What consequences does it have on the cluster health?
+
+    Examples:
+    - Garbage collection in kube-controller-manager will stop working.
+    - Quota will be wrongly computed.
+    - Disabling/removing the CRD is not possible without removing the CR instances. Customer will lose data.
+      Disabling the conversion webhook will break garbage collection.
+
+  - What consequences does it have on existing, running workloads?
+
+    Examples:
+    - New namespaces won't get the finalizer "xyz" and hence might leak resource X
+      when deleted.
+    - SDN pod-to-pod routing will stop updating, potentially breaking pod-to-pod
+      communication after some minutes.
+
+  - What consequences does it have for newly created workloads?
+
+    Examples:
+    - New pods in namespace with Istio support will not get sidecars injected, breaking
+      their networking.
+
+- Does functionality fail gracefully and will work resume when re-enabled without risking
+  consistency?
+
+  Examples:
+  - The mutating admission webhook "xyz" has FailPolicy=Ignore and hence
+    will not block the creation or updates on objects when it fails. When the
+    webhook comes back online, there is a controller reconciling all objects, applying
+    labels that were not applied during admission webhook downtime.
+  - Namespaces deletion will not delete all objects in etcd, leading to zombie
+    objects when another namespace with the same name is created.
+
+-->
+
+## Implementation History
+
+<!--
+Major milestones in the life cycle of a proposal should be tracked in `Implementation
+History`.
+-->
+
+## Alternatives
+
+### Add CSI driver + csi-snapshot-controller creation directly to control-plane-operator
+
+control-plane-operator could directly run the CSI driver and csi-snapshot-controller, without cluster-storage operator,
+cluster-csi-snapshot-controller-operator or aws-ebs-csi-driver-operator.
+
+* Pro: There would be no CRs in the guest cluster, control-plane-operator could read all information it needs from
+  HosterCluster / HostedControlPlane CR.
+* Con: We would end up with two separate code bases that are able to run the operands, and we would need to update both
+  of them when fixing a bug.
+  * Esp. the CSI driver operators are complex, each of them is a bit different and we have ~9 of them (and the number is
+    growing). We would need to reimplement it in the control-plane-operator and the implementation would get out of sync
+    with the standalone OCP easily.
```

Comment:
Could the code to do this be shared between the 2 operators, or would it have to be very different implementations?

Reply:
IMO the implementations would be quite different, because inputs of the operators would be different. In standalone OCP, a CSI driver operator needs to watch Infrastructure, cluster-wide + cloud CA bundles (ConfigMaps), Proxy, ClusterCSIDriver CR and act on their changes. In HyperShift, the input would be only HostedControlPlane CR.

They could share some template of the CSI driver objects, but that's not much.

Reply:
If there's not a lot of code to share and the inputs and outputs are so different, would it make sense to build a special-purpose thing to meet the new requirements? It might prove simpler to maintain than weaving if/then checks throughout the existing code to deal with both cases in one place.